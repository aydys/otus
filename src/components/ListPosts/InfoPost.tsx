import React from 'react'

type Props = {
  posts: any[],
  post: {}
}

export class InfoPost extends React.Component<Props> {
  render() {
    return <div>InfoPost</div>
  }
}