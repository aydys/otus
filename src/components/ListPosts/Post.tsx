import React from "react";

type Props = {
  post: {},
  onClick: (id: number) => void
}

export class Post extends React.Component<Props> {
  render() {
    return <div>Post</div>
  }
}