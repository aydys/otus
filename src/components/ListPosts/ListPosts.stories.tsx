import React from "react";
import { ListPosts } from './ListPosts'
import { withKnobs } from "@storybook/addon-knobs";

export default {
  decorations: [withKnobs],
  title: 'ListPosts'
}

export const ListPostsStory = () => [
  <ListPosts key='jsx' />
]