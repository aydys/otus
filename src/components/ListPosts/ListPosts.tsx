import React from 'react'
import { Post } from "./Post";
import { InfoPost } from "./InfoPost";

import fetch from 'node-fetch'

type PostsProps = {}

interface PostsState {
  posts: any[],
  post: {
    title: string,
    id: number,
    body: string
  }
}

const styled = {
  post: {border: '1px solid grey'},
  containerInfoPost: {width: '40%', float: 'right', marginLeft: 20, border: '1px solid grey'},
  containerList: {width: '50%', float: 'left'}
}

export class ListPosts extends React.Component <PostsProps, PostsState> {

  constructor(props: PostsProps) {
    super(props)
    this.state = {
      posts: [],
      post: {
        title: '',
        id: 0,
        body: ''
      }
    }

    this.getPosts = this.getPosts.bind(this)
    this.onClick = this.onClick.bind(this)
  }

  getPosts<T>(): Promise<T> {
    return fetch("https://jsonplaceholder.typicode.com/posts")
      .then(response => response.json())
  }

  onClick<T>(id: number): void {
    fetch(`https://jsonplaceholder.typicode.com/posts/${id}`)
      .then(response => response.json())
      .then(post => {
        console.log(post)
        this.setState({ post: post })
      } )
  } 

  componentDidMount(): void {
    this.getPosts<object[]>()
      .then(posts => {
        this.setState({ posts })
      })
  }

  render() {
    const { posts, post } = this.state;
    return <React.Fragment>
      <h1>Posts</h1>
      <div>
         <div style={styled.containerList as React.CSSProperties}>
          <ul >
          { posts.map(post => (
              <li key={post.id} style={styled.post}>
                <Post post={post} onClick={this.onClick} />
              </li>
          ))}
        </ul>
        </div>
          
        <div style={styled.containerInfoPost as React.CSSProperties}>
          <InfoPost post={post} />
        </div>
      </div>           
    </React.Fragment>
  }
}
