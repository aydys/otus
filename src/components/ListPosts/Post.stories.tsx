import React from "react";
import { Post } from './Post'
import { object } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";

export default {
  argTypes: { onClick: { action: 'clicked' } },
  title: 'Post'
}

export const ListPost = () => [
  <Post
   post={{title: 'title', body: 'body', id: 34}}
   onClick={action('api calling')}
  />
]