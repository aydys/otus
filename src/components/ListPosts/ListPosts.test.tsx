import React from "react";
import { shallow } from "enzyme";
import { ListPosts } from "./ListPosts";

describe('ListPosts', () => {
  
  const initialState = {
    posts: [
      {
        id: 1,
        title: 'Some Title'
      }
    ],
    post: {
      title: '',
      id: 1,
      body: ''
    }
  }

  const wrapper = shallow(<ListPosts />)
  it('layout dom elements (snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('initial state of component', () => {
    wrapper.setState(initialState);
    expect(wrapper.state()).toEqual(initialState);
  })

  it('exist Post component', () => {
    wrapper.setState(initialState);
    expect(wrapper.find('Post')).toHaveLength(initialState.posts.length);
  })
  it('exist InfoPost component', () => {
    expect(wrapper.find('InfoPost')).toHaveLength(1);
  })
})
