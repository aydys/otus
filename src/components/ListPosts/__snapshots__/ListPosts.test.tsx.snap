// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPosts layout dom elements (snapshot 1`] = `ShallowWrapper {}`;
